## Hamoye_Quiz_Stage_B

### Que 1 Answer - D

### Que 2 Answer - D

### Que 3 Answer - C

### Que 4 Answer - B

### Que 5 Answer - A

### Que 6 Answer - B

### Que 7 Answer - C

### Que 8 Answer - B

### Que 9 Answer - D

### Que 10 Answer - D

### Que 11 Answer - C

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns

import matplotlib.pyplot as plt
import matplotlib.mlab as mlab
import matplotlib
plt.style.use('ggplot')
from matplotlib.pyplot import figure

%matplotlib inline
matplotlib.rcParams['figure.figsize'] = (12,8)

pd.options.mode.chained_assignment = None

pd.get_option("display.max_columns")
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [2]:
data = pd.read_csv('energydata_complete.csv')
data.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4       RH_4         T5   RH_5        T6  \
0  19.79  44.730000  19.000000  45.566667  17.166667  55.20  7.026667   
1  19.79  44.790000  19.000000  45.992500  17.166667  55.20  6.833333   
2  19.79  44.933333  18.926667  45.890000  17.166667  55.09  6.560000   
3  19.79  45.000000  18.890000  45.723333  17.166667  55.09  6.433333   
4  19.79  45.000000  18.890000  45.530000  17.200000  55.09  6.366667   

        RH_6         T7       RH_7    T8       RH_8         T9   RH_9  \
0  84.256667  17.200000  41.626667  18.2  48.900000  17.033333  45.53   
1  84.063333  17.200000  41.560000  18.2  48.863333  17.066667  45.56   
2  83.156667  17.200000  41.433333  18.2  48.730000  17.000000  45.50   
3  83.423333  17.133333  41.290000  18.1  48.590000  17.000000  45.40   
4  84.893333  17.200000  41.230000  18.1  48.590000  17.000000  45.40   

      T_out  Press_mm_hg  RH_out  Windspeed  Visibility  Tdewpoint        rv1  \
0  6.600000        733.5    92.0   7.000000   63.000000        5.3  13.275433   
1  6.483333        733.6    92.0   6.666667   59.166667        5.2  18.606195   
2  6.366667        733.7    92.0   6.333333   55.333333        5.1  28.642668   
3  6.250000        733.8    92.0   6.000000   51.500000        5.0  45.410389   
4  6.133333        733.9    92.0   5.666667   47.666667        4.9  10.084097   

         rv2  
0  13.275433  
1  18.606195  
2  28.642668  
3  45.410389  
4  10.084097

In [3]:
data.shape

(19735, 29)

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19735 entries, 0 to 19734
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         19735 non-null  object 
 1   Appliances   19735 non-null  int64  
 2   lights       19735 non-null  int64  
 3   T1           19735 non-null  float64
 4   RH_1         19735 non-null  float64
 5   T2           19735 non-null  float64
 6   RH_2         19735 non-null  float64
 7   T3           19735 non-null  float64
 8   RH_3         19735 non-null  float64
 9   T4           19735 non-null  float64
 10  RH_4         19735 non-null  float64
 11  T5           19735 non-null  float64
 12  RH_5         19735 non-null  float64
 13  T6           19735 non-null  float64
 14  RH_6         19735 non-null  float64
 15  T7           19735 non-null  float64
 16  RH_7         19735 non-null  float64
 17  T8           19735 non-null  float64
 18  RH_8         19735 non-null  float64
 19  T9  

In [5]:
data.isnull().sum()

date           0
Appliances     0
lights         0
T1             0
RH_1           0
T2             0
RH_2           0
T3             0
RH_3           0
T4             0
RH_4           0
T5             0
RH_5           0
T6             0
RH_6           0
T7             0
RH_7           0
T8             0
RH_8           0
T9             0
RH_9           0
T_out          0
Press_mm_hg    0
RH_out         0
Windspeed      0
Visibility     0
Tdewpoint      0
rv1            0
rv2            0
dtype: int64

In [6]:
# # convert the Purchase Date to datetime
# data['new_date'] = pd.to_datetime(data['date'])
# # add a column for Year
# data['year'] = data['new_date'].dt.year
# data['month'] = data['new_date'].dt.month
# data['day'] = data['new_date'].dt.day
# # print the dataframe
# data.head()

### Checking for % missing values
Yellow color for missing values

Blue color for non-missing values

In [7]:
# if it's a larger dataset and the visualization takes too long can do this.
# % of missing.
for col in data.columns:
#     missing_num = data[col].isnull().sum()
    pct_missing = np.mean(data[col].isnull())
    print('{} - {}%'.format(col, round(pct_missing*100,4)))

date - 0.0%
Appliances - 0.0%
lights - 0.0%
T1 - 0.0%
RH_1 - 0.0%
T2 - 0.0%
RH_2 - 0.0%
T3 - 0.0%
RH_3 - 0.0%
T4 - 0.0%
RH_4 - 0.0%
T5 - 0.0%
RH_5 - 0.0%
T6 - 0.0%
RH_6 - 0.0%
T7 - 0.0%
RH_7 - 0.0%
T8 - 0.0%
RH_8 - 0.0%
T9 - 0.0%
RH_9 - 0.0%
T_out - 0.0%
Press_mm_hg - 0.0%
RH_out - 0.0%
Windspeed - 0.0%
Visibility - 0.0%
Tdewpoint - 0.0%
rv1 - 0.0%
rv2 - 0.0%


### Data statistics and information

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19735 entries, 0 to 19734
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         19735 non-null  object 
 1   Appliances   19735 non-null  int64  
 2   lights       19735 non-null  int64  
 3   T1           19735 non-null  float64
 4   RH_1         19735 non-null  float64
 5   T2           19735 non-null  float64
 6   RH_2         19735 non-null  float64
 7   T3           19735 non-null  float64
 8   RH_3         19735 non-null  float64
 9   T4           19735 non-null  float64
 10  RH_4         19735 non-null  float64
 11  T5           19735 non-null  float64
 12  RH_5         19735 non-null  float64
 13  T6           19735 non-null  float64
 14  RH_6         19735 non-null  float64
 15  T7           19735 non-null  float64
 16  RH_7         19735 non-null  float64
 17  T8           19735 non-null  float64
 18  RH_8         19735 non-null  float64
 19  T9  

In [9]:
data.corr()

Appliances    lights        T1      RH_1        T2      RH_2  \
Appliances     1.000000  0.197278  0.055447  0.086031  0.120073 -0.060465   
lights         0.197278  1.000000 -0.023528  0.106968 -0.005622  0.050985   
T1             0.055447 -0.023528  1.000000  0.164006  0.836834 -0.002509   
RH_1           0.086031  0.106968  0.164006  1.000000  0.269839  0.797535   
T2             0.120073 -0.005622  0.836834  0.269839  1.000000 -0.165610   
RH_2          -0.060465  0.050985 -0.002509  0.797535 -0.165610  1.000000   
T3             0.085060 -0.097393  0.892402  0.253230  0.735245  0.137319   
RH_3           0.036292  0.131161 -0.028550  0.844677  0.121497  0.678326   
T4             0.040281 -0.008859  0.877001  0.106180  0.762066 -0.047304   
RH_4           0.016965  0.114936  0.097861  0.880359  0.231563  0.721435   
T5             0.019760 -0.078745  0.885247  0.205797  0.720550  0.110409   
RH_5           0.006955  0.141233 -0.014782  0.303258  0.029595  0.250271   
T6             0.117638 -0.079029  0.654769  0.316141  0.801186 -0.009670   
RH_6          -0.083178  0.153756 -0.615045  0.245126 -0.580372  0.389933   
T7             0.025801 -0.135347  0.838705  0.021397  0.663660 -0.051422   
RH_7          -0.055642  0.035069  0.135182  0.801122  0.229212  0.690584   
T8             0.039572 -0.071458  0.825413 -0.030053  0.578191 -0.041023   
RH_8          -0.094039  0.012915 -0.006441  0.736196  0.068534  0.679777   
T9             0.010010 -0.157592  0.844777  0.115263  0.675535  0.054544   
RH_9          -0.051462 -0.008766  0.071756  0.764001  0.157346  0.676467   
T_out          0.099155 -0.074424  0.682846  0.340767  0.792255  0.033674   
Press_mm_hg   -0.034885 -0.010576 -0.150574 -0.293957 -0.133028 -0.255646   
RH_out        -0.152282  0.068543 -0.345481  0.274126 -0.505291  0.584911   
Windspeed      0.087122  0.060281 -0.087654  0.204932  0.052495  0.069190   
Visibility     0.000230  0.020038 -0.076210 -0.021057 -0.069721 -0.005368   
Tdewpoint      0.015353 -0.036322  0.571309  0.639106  0.582602  0.499152   
rv1           -0.011145  0.000521 -0.006203 -0.000699 -0.011087  0.006275   
rv2           -0.011145  0.000521 -0.006203 -0.000699 -0.011087  0.006275   

                   T3      RH_3        T4      RH_4        T5      RH_5  \
Appliances   0.085060  0.036292  0.040281  0.016965  0.019760  0.006955   
lights      -0.097393  0.131161 -0.008859  0.114936 -0.078745  0.141233   
T1           0.892402 -0.028550  0.877001  0.097861  0.885247 -0.014782   
RH_1         0.253230  0.844677  0.106180  0.880359  0.205797  0.303258   
T2           0.735245  0.121497  0.762066  0.231563  0.720550  0.029595   
RH_2         0.137319  0.678326 -0.047304  0.721435  0.110409  0.250271   
T3           1.000000 -0.011234  0.852778  0.122737  0.888169 -0.066355   
RH_3        -0.011234  1.000000 -0.140457  0.898978 -0.050062  0.375422   
T4           0.852778 -0.140457  1.000000 -0.048650  0.871813 -0.076489   
RH_4         0.122737  0.898978 -0.048650  1.000000  0.091812  0.352591   
T5           0.888169 -0.050062  0.871813  0.091812  1.000000  0.032786   
RH_5        -0.066355  0.375422 -0.076489  0.352591  0.032786  1.000000   
T6           0.686882  0.076833  0.652350  0.259047  0.629161 -0.078337   
RH_6        -0.647672  0.514912 -0.703149  0.392178 -0.632384  0.263797   
T7           0.847374 -0.250090  0.877763 -0.131204  0.870624 -0.142823   
RH_7         0.172624  0.832685  0.043527  0.894301  0.148905  0.325808   
T8           0.795283 -0.283228  0.796256 -0.167066  0.824981 -0.086991   
RH_8         0.044427  0.828822 -0.095192  0.847259  0.016388  0.359840   
T9           0.901324 -0.195270  0.889439 -0.044518  0.911055 -0.138509   
RH_9         0.134602  0.833538 -0.025549  0.856591  0.072308  0.272197   
T_out        0.699417  0.118207  0.663478  0.293289  0.651321 -0.053126   
Press_mm_hg -0.189974 -0.233274 -0.075292 -0.250748 -0.170999 -0.119327   
RH_out      -0.281718  0.356192 -0.388602  0.336813 -0

## Checking for number of features and observations

In [10]:
data.shape

(19735, 29)

The above result shows we have 19735 rows(observations) and 29 feaures(columns)

### Dropping irrelevant features

In [11]:
data = data.drop(['date', 'lights'], axis = 1)

### Que 12 Solution

In [12]:
# Importing necessary libraries

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

In [13]:
x_T2 = np.array(data['T2'])
y_T6 = np.array(data['T6'])
x_T2 = x_T2.reshape(-1, 1)

train_x, test_x, train_y, test_y = train_test_split(x_T2, y_T6, train_size = 0.7, random_state = 42)

linear = LinearRegression()
linear.fit(train_x, train_y)
prediction_y = linear.predict(test_x)

from sklearn.metrics import r2_score
r_squared = r2_score(test_y, prediction_y)
round(r_squared, 2)

0.64

### Que 13 Solution
Finding the Mean Absolute Error

### Using MinMaxScaler to normalize the data

In [14]:
from sklearn.preprocessing import MinMaxScaler
minMaxScaler = MinMaxScaler()
data_normalize = pd.DataFrame(minMaxScaler.fit_transform(data), columns = data.columns)

In [15]:
data_normalize.head()

Appliances       T1      RH_1        T2      RH_2        T3      RH_3  \
0    0.046729  0.32735  0.566187  0.225345  0.684038  0.215188  0.746066   
1    0.046729  0.32735  0.541326  0.225345  0.682140  0.215188  0.748871   
2    0.037383  0.32735  0.530502  0.225345  0.679445  0.215188  0.755569   
3    0.037383  0.32735  0.524080  0.225345  0.678414  0.215188  0.758685   
4    0.046729  0.32735  0.531419  0.225345  0.676727  0.215188  0.758685   

         T4      RH_4        T5      RH_5        T6      RH_6        T7  \
0  0.351351  0.764262  0.175506  0.381691  0.381070  0.841827  0.170594   
1  0.351351  0.782437  0.175506  0.381691  0.375443  0.839872  0.170594   
2  0.344745  0.778062  0.175506  0.380037  0.367487  0.830704  0.170594   
3  0.341441  0.770949  0.175506  0.380037  0.363800  0.833401  0.164310   
4  0.341441  0.762697  0.178691  0.380037  0.361859  0.848264  0.170594   

       RH_7        T8      RH_8        T9      RH_9     T_out  Press_mm_hg  \
0  0.653428  0.173329  0.661412  0.223032  0.677290  0.372990     0.097674   
1  0.651064  0.173329  0.660155  0.226500  0.678532  0.369239     0.100000   
2  0.646572  0.173329  0.655586  0.219563  0.676049  0.365488     0.102326   
3  0.641489  0.164175  0.650788  0.219563  0.671909  0.361736     0.104651   
4  0.639362  0.164175  0.650788  0.219563  0.671909  0.357985     0.106977   

     RH_out  Windspeed  Visibility  Tdewpoint       rv1       rv2  
0  0.894737   0.500000    0.953846   0.538462  0.265449  0.265449  
1  0.894737   0.476190    0.894872   0.533937  0.372083  0.372083  
2  0.894737   0.452381    0.835897   0.529412  0.572848  0.572848  
3  0.894737   0.428571    0.776923   0.524887  0.908261  0.908261  
4  0.894737   0.404762    0.717949   0.520362  0.201611  0.201611

In [16]:
data_normalize.tail()

Appliances        T1      RH_1        T2      RH_2        T3      RH_3  \
19730    0.084112  0.926786  0.537657  0.711655  0.606309  0.830841  0.579374   
19731    0.074766  0.919747  0.536006  0.701769  0.607836  0.825302  0.582178   
19732    0.242991  0.919747  0.538666  0.692651  0.627198  0.818378  0.603988   
19733    0.383178  0.919747  0.549491  0.677054  0.634717  0.805085  0.585294   
19734    0.392523  0.919747  0.538758  0.666171  0.632902  0.799546  0.579062   

             T4      RH_4        T5      RH_5        T6  RH_6        T7  \
19730  0.864865  0.765258  0.752031  0.339590  0.898317   0.0  0.858624   
19731  0.864865  0.765258  0.754897  0.338487  0.880852   0.0  0.864010   
19732  0.864865  0.771233  0.754897  0.337585  0.864260   0.0  0.862394   
19733  0.864865  0.773794  0.752031  0.336583  0.829525   0.0  0.858624   
19734  0.864865  0.781192  0.752031  0.336583  0.788580   0.0  0.858624   

           RH_7        T8      RH_8        T9      RH_9     T_out  \
19730  0.755319  0.768386  0.701645  0.864724  0.729443  0.891747   
19731  0.752280  0.768386  0.691912  0.864724  0.729443  0.887460   
19732  0.751773  0.768386  0.687457  0.864724  0.729443  0.883173   
19733  0.748075  0.764953  0.682617  0.864724  0.730581  0.878885   
19734  0.739504  0.771681  0.690062  0.864724  0.731719  0.874598   

       Press_mm_hg    RH_out  Windspeed  Visibility  Tdewpoint       rv1  \
19730     0.602326  0.416667   0.238095    0.348718   0.901961  0.861981   
19731     0.602326  0.421053   0.250000    0.361538   0.900452  0.985726   
19732     0.602326  0.425439   0.261905    0.374359   0.898944  0.583979   
19733     0.602326  0.429825   0.273810    0.387179   0.897436  0.126371   
19734     0.602326  0.434211   0.285714    0.400000   0.895928  0.682391   

            rv2  
19730  0.861981  
19731  0.985726  
19732  0.583979  
19733  0.126371  
19734  0.682391

### Separating the dataset into predictive features and target
Taking Appliance as the target feature

In [17]:
X = data_normalize.drop(['Appliances'], axis = 1)
y = data_normalize['Appliances']

### Splitting the dataset into train and test sets using train_test_split

In [18]:
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size = 0.7, random_state = 42)

In [19]:
linear = LinearRegression()
linear.fit(X_train, y_train)
y_prediction = linear.predict(X_test)

In [20]:
from sklearn.metrics import mean_absolute_error

mae = mean_absolute_error(y_test, y_prediction)
print(round(mae, 2))

0.05


### Que 14 Solution

In [21]:
## Residual Sum of Squares(RSS)

residual_ss = np.sum(np.square(y_test - y_prediction))
print(round(residual_ss, 2))

45.35


### Que 15 Solution

In [22]:
## Root Mean Squared Error(RMSE)

from sklearn.metrics import mean_squared_error
rmse = np.sqrt(mean_squared_error(y_test, y_prediction))
print(round(rmse, 3))

0.088


### Que 16 Solution

In [23]:
from sklearn.metrics import r2_score
r_squared = r2_score(y_test, y_prediction)
round(r_squared, 2)

0.15

### Que 17 Solution

In [24]:
#to check the weights of the features using linear regression

def weights(model, feature, column):
    #this function returns the weight of every feature
    weights = pd.Series(linear.coef_, feature.columns).sort_values()
    weights_df = pd.DataFrame(weights).reset_index()
    weights_df.columns = ['Features', column]
    weights_df[column].round(3)
    return weights_df

linear_weights = weights(linear, X_train, 'Linear_model_Weight')
linear_weights

Features  Linear_model_Weight
0          RH_2            -0.456698
1         T_out            -0.321860
2            T2            -0.236178
3            T9            -0.189941
4          RH_8            -0.157595
5        RH_out            -0.077671
6          RH_7            -0.044614
7          RH_9            -0.039800
8            T5            -0.015657
9            T1            -0.003281
10          rv1             0.000770
11          rv2             0.000770
12  Press_mm_hg             0.006839
13           T7             0.010319
14   Visibility             0.012307
15         RH_5             0.016006
16         RH_4             0.026386
17           T4             0.028981
18    Windspeed             0.029183
19         RH_6             0.038049
20         RH_3             0.096048
21           T8             0.101995
22    Tdewpoint             0.117758
23           T6             0.236425
24           T3             0.290627
25         RH_1             0.553547

In [25]:
print(linear_weights['Linear_model_Weight'].min())
print(linear_weights['Linear_model_Weight'].max())

-0.4566979483385004
0.5535465998386391


### Que 18 Solution

In [26]:
## ridge regression with alpha = 0.4
from sklearn.linear_model import Ridge

ridge = Ridge(alpha=0.4)
ridge.fit(X_train, y_train)
ridge_pred = ridge.predict(X_test)
train_score = ridge.score(X_train, y_train)
test_score = ridge.score(X_test, y_test)
# print("Training score {:.3f}".format(train_score))
# print("Testing score {:.3f}".format(test_score))

## Root Mean Squared Error(RMSE) for ridge model

from sklearn.metrics import mean_squared_error
rmse = np.sqrt(mean_squared_error(y_test, ridge_pred))
print(round(rmse, 3))

0.088


### Que 19 Solution

In [27]:
## lasso model with alpha = 0.001

from sklearn.linear_model import Lasso

lasso = Lasso(alpha = 0.001)
lasso.fit(X_train, y_train)
lasso_pred = lasso.predict(X_test)
# train_score = lasso.score(X_train, y_train)
# test_score = lasso.score(X_test, y_test)
# print("{:.3f}".format(train_score))
# print("{:.3f}".format(test_score))
print("Number of features with non-zero weights for training: {}".format(np.sum(lasso.coef_ != 0)))

Number of features with non-zero weights for training: 4


### Que 20 Solution

In [28]:
## Root Mean Squared Error(RMSE) with lasso regression

from sklearn.metrics import mean_squared_error
rmse = np.sqrt(mean_squared_error(y_test, lasso_pred))
print('Root Mean Squared Error(RMSE) with lasso regression is {}'.format(round(rmse, 3)))

Root Mean Squared Error(RMSE) with lasso regression is 0.094
